### ISSUE TITLE
XUnit "Timeout" parameter on Fact attribute does not work with synchronous test/code

### ISSUE THREAD

**Issue by mcallaghan-bsm** [xunit/xunit#2222]
As per the history of #217, we are aware of an issue with the `Fact(Timeout)` attribute/parameter not working in synchronous test code, however it does if developers take care to only use asynchronous.

``` c#
[Fact(Timeout = 1000)]
```

First to start with some references, it seems this parameter is not actually officially documented.
If we look to #1900 we can see why.  However https://xunit.net/docs/comparisons does not list the `Fact(Timeout)` attribute/parameter either.  So from the start, it is not clear to me if this is formally a supported attribute or not.  None-the-less, compared to NUnit, I suppose we could assert that the community is roughly in agreement that this functionality / capability is necessary for reliable CI and tests.

The REQUIREMENT:
* When running tests with `XUnit`, if a test exceeds `## seconds`, kill it (timed out), "aka took too long"
 ^^ this is to protect CI or local runners from running excessively long in a deadlock/hang/starvation/etc type situation

To clearly summarise what works and what does not:

(snip from: https://github.com/xunit/xunit/issues/217#issuecomment-625402054)

There is a current bug with Timeout that it only works with async tests.

So this works:
``` c#
[Fact(Timeout = 1000)]
public async Task Passing()
{
    await Task.Delay(5000);
}
```
but this does not:
``` c#
[Fact(Timeout = 1000)]
public void Failing()
{
    Thread.Sleep(5000);
}
```

This is important for developers to know in order to workaround the gap, but ideally XUnit might be able to fix it.

One example situation however where it is impossible to workaround is if the test code is leveraging other libraries (in source or upstream) - that are synchronous ... we may have our hands tied here, and any `Thread.Sleep()` elsewhere in the code under test or supporting libraries can break our ability to rely on XUnit's `Fact(Timeout)`.

Further, this is in direct compare to: https://docs.nunit.org/articles/nunit/writing-tests/attributes/timeout.html , which as far as it has been stated by others in #217, works as documented without a hitch for synchronous and asynchronous code.

**Comment by gabecook** [comment]
Thank you for your work on XUnit!

I'm not sure if this is the right place for this, but here is what I ended up with as a workaround. Are there any gotchas with XUnit?

```C#
    /// <summary>
    /// Helper static methods for queuing specified work on the thread pool using Task.Run() but with a timeout. 
    /// </summary>
    public static class TaskWithTimeout
    {
        public const int DefaultTimeout = 10_000;

        /// <summary>
        /// Queues the specified work to run on the thread pool and returns a Task object that represents that work.
        /// </summary>
        /// <param name="action">The work to execute asynchronously</param>
        /// <param name="exceptionMessage">The optional message for the TimeoutException.</param>
        /// <returns>A task that represents the work queued to execute in the ThreadPool.</returns>
        /// <exception cref="TimeoutException">The action takes longer than the DefaultTimeout value.</exception>
        public static  async Task Run(Action action, string exceptionMessage = null)
        {
            await Run(DefaultTimeout, action, exceptionMessage);
        }

        /// <summary>
        /// Queues the specified work to run on the thread pool and returns a Task object that represents that work.
        /// </summary>
        /// <param name="timeout">The timeout value.</param>
        /// <param name="action">The work to execute asynchronously</param>
        /// <param name="exceptionMessage">The optional message for the TimeoutException.</param>
        /// <returns>A task that represents the work queued to execute in the ThreadPool.</returns>
        /// <exception cref="TimeoutException">The action takes longer than the specified timeout value.</exception>
        public static async Task Run(int timeout, Action action, string exceptionMessage = null)
        {
            var task = Task.Run(action);
            if (task.Wait(timeout))
            {
                await task;
            }
            else
            {
                if (exceptionMessage != null)
                {
                    throw new TimeoutException(exceptionMessage);
                }
                throw new TimeoutException();
            }
        }

        /// <summary>
        /// Queues the specified work to run on the thread pool and returns a Task object that represents that work.
        /// A cancellation token allows the work to be cancelled if it has not yet started.
        /// </summary>
        /// <param name="action">The work to execute asynchronously</param>
        /// <param name="token">A cancellation token that can be used to cancel the work if it has not yet started. Does not pass cancellationToken to action.</param>
        /// <param name="exceptionMessage">The optional message for the TimeoutException.</param>
        /// <returns>A task that represents the work queued to execute in the ThreadPool.</returns>
        /// <exception cref="TimeoutException">The action takes longer than the DefaultTimeout value.</exception>
        public static  async Task Run(Action action, CancellationToken token, string exceptionMessage = null )
        {
            await Run(DefaultTimeout, action, token, exceptionMessage);
        }

        /// <summary>
        /// Queues the specified work to run on the thread pool and returns a Task object that represents that work.
        /// A cancellation token allows the work to be cancelled if it has not yet started.
        /// </summary>
        /// <param name="timeout">The timeout value.</param>
        /// <param name="action">The work to execute asynchronously</param>
        /// <param name="token">A cancellation token that can be used to cancel the work if it has not yet started. Does not pass cancellationToken to action.</param>
        /// <param name="exceptionMessage">The optional message for the TimeoutException.</param>
        /// <returns>A task that represents the work queued to execute in the ThreadPool.</returns>
        /// <exception cref="TimeoutException">The action takes longer than the specified timeout value.</exception>
        public static async Task Run(int timeout, Action action, CancellationToken token, string exceptionMessage = null)
        {
            var task = Task.Run(action, token);
            if (task.Wait(timeout))
            {
                await task;
            }
            else
            {
                if (exceptionMessage != null)
                {
                    throw new TimeoutException(exceptionMessage);
                }
                throw new TimeoutException();
            }
        }
    }
}
```

Usage is straightforward and explicit.

```C#
public class ExampleTest
{
     [Fact]
     public async Task TestTest()
     {
            await TaskWithTimeout.Run(() =>
            {
                 .... unstable C interop code that might never return ...

                Assert.True(false);
            });
     }
}

**Comment by jefflill** [comment]
I wonder if perhaps it would be easier if we redefined that **timeout** means here, changing this from trying to accurately enforce these timeouts to just _trying_ to **prevent a test case from running forever**.

How about something like this:

* Xunit retains the async-only test case limitation 
* Xunit starts a `Stopwatch` when the test starts
* Xunit periodically compares the elapsed stopwatch time against the timeout
* Tests running longer than the timeout are aborted as you do now
* Parallel tests could be supported but folks would need to recognize that the timeout may need to be lengthened to account for the extra work performed by other tests

The nice thing about using the `Stopwatch` is that this will also include any time spent in `Thread.Sleep()` when detecting when the test has run too long.

The downsides are:

* Synchronous tests are still not supported
* It probably won't be possible to actually kill a thread executing a long `Thread.Sleep()` call (no `Thread.Abort()` on .NET Core)

This could be targeted at **integration test scenarios** where the tests are more likely to run for longer and there's more chance of transient issues (e.g. slow network) impacting things.

So, the idea would be to specify relatively long timeouts for integration tests.  So for a test that will take maybe 10 seconds under ideal conditions, we'd set a timeout of perhaps 2 minutes to account for time spent running other parallel tests as well as any transient issues.  This would help prevent the test from running much longer than 2 minutes (hoping that nobody upstream is doing something really stupid with `Thread.Sleep()`).

**Comment by jefflill** [comment]
In addition, it would be nice if the test runner could be configured somehow to **limit how long all tests in a project can run collectively**, perhaps via an assembly attribute.

The idea here is that when this limit has been exceeded for a test project, then the runner will fail all pending tests with an `TestProjectTimeoutException` (or something).  Then developers could configure reasonable values this to help prevent tests from running longer than the GitHub job.

**Comment by nohwnd** [comment]
> Further, this is in direct compare to NUnit: https://docs.nunit.org/articles/nunit/writing-tests/attributes/timeout.html , which as far as it has been stated by others in #217, works as documented without a hitch for synchronous and asynchronous code.

It does not work unfortunately, it only seems like it. When timeout is reached it simply completes the task, reporting the failure. But the code is running on the background until it finishes, or until all tests in that assembly are finished and the process is killed. 

On .NET "Core" there is no api to abort a thread, which is why we kill the process on timeout in MSTest. It is overkill, but it is the only thing we can do to stop the code from running without being able to abort the thread. 

If you write NUnit test like this: 

```cs
using NUnit.Framework;
using System.IO;

namespace TestProject10
{
    public class Tests
    {
        [SetUp]
        public void Setup()
        {
            File.AppendAllLines(@"C:\t\nunit.txt", new[] { "--setup--" });
        }

        [TearDown]
        public void Teardown()
        {
            File.AppendAllLines(@"C:\t\nunit.txt", new[] { "--teardown--" });
        }

        [Test]
        [Timeout(1000)]
        public void Test1()
        {
            System.Threading.Thread.Sleep(1200);
            File.AppendAllLines(@"C:\t\nunit.txt", new[] { "hello" });
            System.Threading.Thread.Sleep(1000);
            File.AppendAllLines(@"C:\t\nunit.txt", new[] { "hello hello" });
        }
    }

    public class Tests2
    {
        [Test]
        public void Test2()
        {
            System.Threading.Thread.Sleep(2000);
            File.AppendAllLines(@"C:\t\nunit.txt", new[] { "hello2" });
        }
    }
}
```

This is the output you will see: 

```
 gc C:\t\nunit.txt -Wait -tail 0
--setup--
hello
hello hello
hello2
--setup--
hello
hello hello
hello2
```

Test 1 will report that it failed after 1s, but it will happily finish running the code on the background while test2 is running.

Also teardown won't be executed. Which is curse and a blessing at the same time. Curse because you are cleaning up in it. But blessing because your teardown won't start fighting with the supposedly cancelled test for the resource that the test is holding, and the teardown is supposed to clean up. 

To solve all this test authors would need to start: 
- writing code that can cancel itself correctly, because cancellation is not forcing the code to abort itself, it just raises hand so it knows cancellation was requested, but if the code ignores it, nothing happens
- write thread safe code for the teardown, because the main code might still be running

- OR - 

Add Thread.Abort back to .NET.


More information about how it works in MSTest here: https://github.com/microsoft/testfx/issues/522#issuecomment-567512940

**Comment by draganjovanovic1** [comment]
My two cents for F# workaround. 

I have test like this which started failing with error `Tests marked with Timeout are only supported for async tests` once I updated `xunit` to 2.7:

```fsharp
[<Fact(Timeout=300_000)>]
member _.``Recurring waypoints multiple visits`` () =
    task {
        // async test body
    }
```

Workaround is to upcast to `Task` like this as `xunit` for some reason doesn't like `Task<unit>` (which is how F# compiler infers return type of the test method):

```fsharp
    task {
        // async test body
    } :> Task
```

**Comment by bradwilson** [comment]
@draganjovanovic1 Thanks, I've opened an issue to get this resolved in the next build.

**Comment by naasking** [comment]
Even if a timeout that kills a test isn't feasible, at the very least have it output a message that it's exceeding some maximum  execution time so we know which tests are causing the test suite to hang. The VS runner has LongRunningTestSeconds and we can at least inspect it which tests are still running, but the console runner on a CI machine is kind of a black box and debugging these hangs is very tedious.

**Comment by bradwilson** [comment]
> The VS runner has LongRunningTestSeconds and we can at least inspect it which tests are still running, but the console runner on a CI machine is kind of a black box and debugging these hangs is very tedious.

Which console runner? LongRunningTestSeconds should work in every runner, assuming you've also enabled DiagnosticMessages. If you're using `dotnet test`, it tries to hide all of our output by default, so you can add `--logger "console;verbosity=normal"` so you can see our output (including diagnostic messages like tests that are running too long).

**Comment by bradwilson** [comment]
> So basically in the console output I should look for is "[Long Running Test]".

Correct.

**Comment by bradwilson** [comment]
@nohwnd wrote:

> On .NET "Core" there is no api to abort a thread, which is why we kill the process on timeout in MSTest. It is overkill, but it is the only thing we can do to stop the code from running without being able to abort the thread.

Right now in v2, we're in the same boat as NUnit in that async tests will still continue to run even after we've timed them out. We do not provide the developer with access to any cancellation token that they could wait on, and there's no way to forcefully abort a `Task`.

For v3, I've been thinking about this (after verifying that `Thread.Abort()` is still not available as of .NET 8), and I'm torn. We provide the developer with access to a cancellation token via `TestContext.Current.CancellationToken`. Whether the test is an `async` method or not, they could be using this cancellation token to know when it's time to stop running, which means we should be able to support Timeout in v3 even with non-`async` test methods.

I'm tempted to say that after we time out the test, we give them some small amount of wait time before we just kill the test process entirely. How have users reacted to this behavior from MSTest? Do you also give them an amount of time to wait before terminating the process? If so, how much, and can they configure that?

**Comment by nohwnd** [comment]
We have reverted to the task based approach without killing the process, and are getting closer and closer to having a usable cooperative cancellation.

Our users rarely run into situations where continuing the test on the background causes problems, and in the past we had way more issues filed about the process terminating and not continuing to other tests, especially when there were tests running in parallel.

I would also prefer a way to abort the thread, no matter if it remains in unknown state.

**Comment by JohnLeyva** [comment]
If understand correctly TimeOut purpose are
(a) to assert the test case does not takes longer than expected 
(b) to stop execution as soon is is clear that the assertion is not longer possible to succeed.

From my POV (a) is the main responsibility and (b) is an optimization.

Lets say  timeOut=1000   (1s)  and the test takes 1hr to run, is ok to run for 1 hr, but the test should be marked as failed.

Here the purpose is to alert a developer that last commit changes, does not fit expectations and ASAP should review and fix it.

Certainly, providing (b) is a important optimization, and the testing framework (xunit) should just try to do its best effort.

**Comment by bradwilson** [comment]
@JohnLeyva wrote:

> If I understand correctly TimeOut purposes are
> (a) to assert the test case does not take longer than expected
> (b) to stop execution as soon is clear that the assertion is not longer possible to succeed.

Yes, I'd say your assessment is correct: the primary value is to assert that the test does not run overlong.

However, it's important to some people that a test which has failed should stop running; it normally does when failure is because of an assertion failure or a thrown exception, but it doesn't when it fails because of running overlong. For some, wasting time & resources is a very bad thing, as a build host with a test running an hour that failed a minute in is a build host that could've/should've been running the next scheduled build (for the same project or another one). Worse, you may be running in a build environment where you don't get immediate notification of build failures, so you could be waiting that entire hour for the notification that the build _has_ failed, rather than knowing one second in that the build _will eventually_ fail.

So yes, it's an optimization, but not a purposeless one in my opinion.

I talked above about adding an analyzer related to doing Timeout without `CancellationToken`, which I will open right now (not sure why I didn't in the first place). I have already added an analyzer which will flag in v3 tests if you're calling a method that accepts `CancellationToken` and you aren't passing one to it: https://github.com/xunit/xunit.analyzers/commit/2188a2df0d7ad684dd10fefdd250973b998b5aee

**Comment by naasking** [comment]
> For some, wasting time & resources is a very bad thing, as a build host with a test running an hour that failed a minute in is a build host that could've/should've been running the next scheduled build (for the same project or another one).

Even worse if it runs afoul of some sort of deadlock, and then the test never actually ends at all. I've run into while testing using third-party dependencies, like Selenium.

**Comment by bradwilson** [comment]
Agreed, which is why most CI servers can set an overall build timeout.

**Comment by bradwilson** [comment]
Available in v3 `0.3.0-pre.22` https://xunit.net/docs/using-ci-builds
